### ISSUE TITLE
No metrics for etcd and kube-proxy

### ISSUE THREAD

**Issue by shibumi** [kubermatic/kubeone#1215]
Hi,
first of all thank you for kubeone. It works really good and my company is more and more thinking about using it in production.
I just have a few more issues with it. One of these issues is monitoring etcd and kube-proxy components with prometheus-stack.

```
❯ k get po -n kube-system
NAME                                             READY   STATUS    RESTARTS   AGE
calico-kube-controllers-8585f5b857-5bg2d         1/1     Running   1          52m
canal-4v76p                                      2/2     Running   2          7d6h
canal-c92hr                                      2/2     Running   2          7d6h
canal-nmpvd                                      2/2     Running   8          7d7h
canal-qmjjk                                      2/2     Running   7          7d7h
canal-v2sq6                                      2/2     Running   6          7d7h
canal-zlg2z                                      2/2     Running   2          7d6h
coredns-66bff467f8-92kg5                         1/1     Running   2          7d7h
coredns-66bff467f8-xlqcg                         1/1     Running   0          42m
etcd-k8scontrol1.avency.com                      1/1     Running   4          7d7h
etcd-k8scontrol2.avency.com                      1/1     Running   5          7d7h
etcd-k8scontrol3.avency.com                      1/1     Running   2          7d7h
kube-apiserver-k8scontrol1.avency.com            1/1     Running   5          7d7h
kube-apiserver-k8scontrol2.avency.com            1/1     Running   4          7d7h
kube-apiserver-k8scontrol3.avency.com            1/1     Running   3          7d7h
kube-controller-manager-k8scontrol1.avency.com   1/1     Running   5          7d7h
kube-controller-manager-k8scontrol2.avency.com   1/1     Running   5          7d7h
kube-controller-manager-k8scontrol3.avency.com   1/1     Running   4          7d7h
kube-proxy-bvwbr                                 1/1     Running   2          7d7h
kube-proxy-hc5cz                                 1/1     Running   4          7d7h
kube-proxy-hkl4m                                 1/1     Running   1          7d6h
kube-proxy-lnz44                                 1/1     Running   1          7d6h
kube-proxy-nlvx6                                 1/1     Running   3          7d7h
kube-proxy-zffmv                                 1/1     Running   1          7d6h
kube-scheduler-k8scontrol1.avency.com            1/1     Running   5          7d7h
kube-scheduler-k8scontrol2.avency.com            1/1     Running   4          7d7h
kube-scheduler-k8scontrol3.avency.com            1/1     Running   5          7d7h
metrics-server-7ff9c6f79c-vzkln                  1/1     Running   3          7d7h
node-local-dns-6bgc8                             1/1     Running   1          7d6h
node-local-dns-78vg9                             1/1     Running   2          7d7h
node-local-dns-brpt8                             1/1     Running   2          7d6h
node-local-dns-fdn8s                             1/1     Running   5          7d7h
node-local-dns-vcdz8                             1/1     Running   1          7d6h
node-local-dns-xsn4h                             1/1     Running   3          7d7h
sealed-secrets-controller-55d459cc4c-xwqp2       1/1     Running   0          27h
```

The metrics do work for everything, except for etcd and kube-proxy. Now I am wondering. Is this an issue with kubeone and kubeone's way to  bootstrap it? Or is this a misconfiguration on my side?

The exact error message in the prometheus UI for etcd is:
```
Get "http://10.88.10.2:2379/metrics": read tcp 10.244.3.57:41398->10.88.10.2:2379: read: connection reset by peer
```
For kube-proxy I get this message here:
```
Get "http://10.88.10.11:10249/metrics": dial tcp 10.88.10.11:10249: connect: connection refused
```

Other components seem to work without issues. When I have a look into the configuration for etcd I see this:
```
    - --listen-metrics-urls=http://127.0.0.1:2381
```

This would explain that the metric is not reachable from outside of the node. The metrics-url for kube-proxy is missing at all in the configuration, too. That's why I think this is maybe a misconfiguration on side of kubeone. How am I supposed to  monitor these components?

**Comment by kron4eg** [comment]
That's the way kubeadm initialized those components, we don't configure anything special here.

**Comment by shibumi** [comment]
@kron4eg can I modify the kubeadm configuration on a live cluster? If no, shouldn't kubeone provide an option to configure kubeadm flags?

**Comment by shibumi** [comment]
I had a fast glimpse on the kubeone sourcecode but I couldn't find the part that bootstraps the etcd cluster..if iirc we just need to pass a configuration flag from the kubeone.yaml into the code that bootstraps the cluster.

**Comment by kron4eg** [comment]
OK, so two points:
* kube-proxy is configured from the kube-proxy configMap and it's possible to reconfigure it to your needs after cluster has been created
* etcd's `--listen-metrics-urls` is kinda misleading, as it's possible to obtain etcd metrics from it's regular client endpoints, you only have to use TLS credentials (see https://github.com/kubernetes/kubernetes/pull/90954#discussion_r422998704)

**Comment by shibumi** [comment]
@kron4eg  Thanks. I think this should fix the issue for me.

**Comment by shibumi** [comment]
Hey @kron4eg  sorry for updating this issue:

I have updated to kubernetes 1.19 and it seems that kube-scheduler and kube-controller-manager are both configured for 127.0.0.1 only and therefore I am not able to monitro them with prometheus.

Do you have any idea? Will there be an option to configure them in the future in kubeone?

**Comment by kron4eg** [comment]
@shibumi Here's what community has to say about this: https://github.com/prometheus-operator/kube-prometheus/issues/913#issuecomment-776369401

**Comment by kron4eg** [comment]
As such proxy mentioned in previous link I can suggest https://github.com/brancz/kube-rbac-proxy

**Comment by shibumi** [comment]
@kron4eg  Thanks a lot for your input! Very useful.

**Comment by randrusiak** [comment]
@shibumi Could you share how did you resolve this problem?

**Comment by shibumi** [comment]
@randrusiak sure.

For the kubeProxy I have edited the the configmap `kube-proxy` in the `kube-system` namespace. There is a variable called `metricsBindAddress`. This address is usually set to 127.0.0.1. I fixed this via setting the address `0.0.0.0` (**ATTENTION: Setting 0.0.0.0 might expose the kubeProxy metrics to the internet. Make sure you have a firewall**).

The etcd configuration has been a little bit more difficult. There you have to export the etcd certificates, ca and key first:
```
$ kubectl exec etcd-k8scontrol1  -it -- cat /etc/kubernetes/pki/etcd/ca.crt > ca.crt
$ kubectl exec etcd-k8scontrol1  -it -- cat /etc/kubernetes/pki/etcd/healthcheck-client.crt > healthcheck-client.crt
$ kubectl exec etcd-k8scontrol1  -it -- cat /etc/kubernetes/pki/etcd/healthcheck-client.key > healthcheck-client.key
```

Then you have to create a secret in the monitoring namespace for this:
```
$ kubectl create secret generic etcd-certs -n monitoring --from-literal=ca.crt="$(cat ca.crt)" --from-literal=healthcheck- \ 
  client.crt="$(cat healtcheck-client.crt)" --from-literal=healthcheck-client.key="$(cat healtcheck-client.key)"
```

And then you have to configure the kube-prometheus helm chart properly to use the correct etcd port (2380) + the existing etcd-certs secret.

Why do you need to do this? If etcd runs in 'secure mode' it encrypts all traffic, hence the etcd pods are only reachable with a valid certificate. That's why you have to attach the etcd ca, cert and key to all monitoring requests.

**Comment by randrusiak** [comment]
@shibumi Thank you for the detailed explanation. I'm wondering how do you get access to kube-controller-manager and kube-scheduler?
Do you use a proxy or decided to expose ports publicly as well?

**Comment by shibumi** [comment]
@randrusiak oh, this is a good question. I just disabled monitoring for both of them, because I couldn't find a proper answer for these two components.

You can modify the kubelet-config, but this is something more complicated: https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/

**Comment by randrusiak** [comment]
I rather avoid making manual changes so I decided to test the solution with haproxy and it seems to work great. 
I also added config for kube-proxy therefore there is no need to exposing anything publicly. 

Thank you once again for your answers it helped me a lot :)

**Comment by shibumi** [comment]
@randrusiak How do you monitor the etcd setup?

**Comment by randrusiak** [comment]
I monitor etcd in the same way as you describe above.

**Comment by c4tz** [comment]
`kube-etcd` seems to provide a `/metrics` endpoint on port `2381` now, thus I was able to monitor it without having to create a secret containing the certificates. All one has to do (for the people coming here from Google, like me) is to create a daemon-set running the aforementioned [ha-proxy](https://github.com/prometheus-operator/kube-prometheus/issues/913#issuecomment-776369401), which can be included in KubeOne as an addon:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: metrics-proxy-config
  namespace: kube-system
data:
  haproxy.cfg: |
    defaults
      mode http
      timeout connect 5000ms
      timeout client 5000ms
      timeout server 5000ms
      default-server maxconn 10

    frontend kube-controller-manager
      bind ${NODE_IP}:10252
      http-request deny if !{ path /metrics }
      default_backend kube-controller-manager
    backend kube-controller-manager
      server kube-controller-manager 127.0.0.1:10257 ssl verify none

    frontend kube-scheduler
      bind ${NODE_IP}:10251
      http-request deny if !{ path /metrics }
      default_backend kube-scheduler
    backend kube-scheduler
      server kube-scheduler 127.0.0.1:10259 ssl verify none

    frontend kube-proxy
      bind ${NODE_IP}:10249
      http-request deny if !{ path /metrics }
      default_backend kube-proxy
    backend kube-proxy
      server kube-proxy 127.0.0.1:10249

    frontend kube-etcd
      bind ${NODE_IP}:2381
      http-request deny if !{ path /metrics }
      default_backend kube-etcd
    backend kube-etcd
      server kube-etcd 127.0.0.1:2381
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: metrics-proxy
  namespace: kube-system
spec:
  selector:
    matchLabels:
      name: metrics-proxy
  template:
    metadata:
      labels:
        name: metrics-proxy
    spec:
      hostNetwork: true
      tolerations:
      - key: node-role.kubernetes.io/master
        operator: Exists
        effect: NoSchedule
      containers:
      - name: metrics-proxy
        image: haproxy:2.5-alpine
        resources:
          limits:
            memory: 200Mi
          requests:
            cpu: 50m
            memory: 175Mi
        volumeMounts:
        - name: config
          mountPath: /usr/local/etc/haproxy
        env:
        - name: NODE_IP
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: status.hostIP
      terminationGracePeriodSeconds: 30
      volumes:
        - name: config
          configMap:
            name: metrics-proxy-config

```
and point Prometheus to the right port:
```yaml
kubeEtcd:
  service:
    port: 2381
    targetPort: 2381
```
This will enable Prometheus to retrieve metrics from `kube-controller-manager`, `kube-scheduler` and `kube-proxy` as well as `kube-etcd`.

**Comment by dirien** [comment]
hi @c4tz, thanks for the post! It helped me with our current setting!

**Comment by sebastiangaiser** [comment]
@c4tz how about the `kube-controller-manager` and `kube-scheduler`. Have you configured them for prometheus?

**Comment by c4tz** [comment]
AFAIR, I didn't (have to) configure anything other than the things mentioned in my comment above.

**Comment by sebastiangaiser** [comment]
@c4tz thanks for your reply. We tried the exact same using latest kube-prometheus-stack and we didn't get any metrics for `kube-controller-manager` and `kube-scheduler`. `kube-proxy` and `kube-etcd` are working fine with your solution.

@kron4eg do you have some input, maybe for debugging this? We are running on Kubernetes `1.23.4`.

**Comment by baryonicnonsense** [comment]
For those that stumble upon this, kube-prometheus-stack defaults to using HTTPS to scrape the kube-controller-manager and kube-scheduler metrics endpoints. Setting
```
kubeControllerManager:
  serviceMonitor:
    https: false
kubeScheduler:
  serviceMonitor:
    https: false
```
in values.yaml will allow Prometheus to correctly scrape the endpoints exposed by the HAProxy DaemonSet mentioned in https://github.com/kubermatic/kubeone/issues/1215#issuecomment-992471229.

**Comment by jgagnon44** [comment]
@c4tz I tried your suggestion and it does work for anything not running on a master node. For example, `kube-proxy` runs on all nodes. When I put this in place, I find that Prometheus is able to access `kube-proxy` on all of the _worker_ nodes, but not on the master. `kube-controller-manager`, `kube-etcd` and `kube-scheduler` are running on the master node _only_ and still cannot be accessed. BTW I'm using Kubernetes 1.24.

Does anyone have an idea what is special about the master node that prevents these services from being accessed? Does anyone have ideas about how this can be accomplished?

As a side note, I was able to access `kube-controller-manager` and `kube-scheduler` from Prometheus without this proxy technique, but only by directly modifying the config of the running pods on the master node. Certainly not a recommended practice.
